### PR TITLE
Refactoring blocking closer to chan

### PR DESCRIPTION
**Pull request by bfosberry**
High level overview of this change is that instead of dumping everything into a buffer as we receive it, we're using a channel. This channel is not buffered on overall byte size, but rather number of messages, so be aware of this variation when setting limits.

The new behaviour is that writing will fill the channel to a point, after which writing will block. Reading will use a buffer to aggregate channel data to fill the appropriate requested byte array, and will read using a timeout from the channel.

If no data is available the read will block. (still need to add this bit)

### CODE REVIEW

**Comment by ngauthier** [comment]
Nice!

**Comment by ngauthier** [comment]
Actually I think this is just a pipe:

https://godoc.org/io#Pipe

**Comment by bfosberry** [comment]
BOOM

Only variance from the behavior of Pipe is that should the stream be closed we can still read the last nugget of data before closing the reader out

File: blocking_read_write_closer.go [hunk 1]
```diff
@@ -14,54 +11,31 @@ func NewBlockingReadWriteCloser() io.ReadWriteCloser {
...
 type blockingReadWriteCloser struct {
-	buffer *bytes.Buffer
-	cv     *sync.Cond
-	closed VolatileBool
+	reader *io.PipeReader
+	writer *io.PipeWriter
+	closed bool
 }
 
 func newBlockingReadWriteCloser() *blockingReadWriteCloser {
```
Comment (ngauthier) [inline blocking_read_write_closer.go]: can we move this into the public constructor? We don't need the private constructor.

**Comment by ngauthier** [comment]
Ah, because normally you'd close the sides separately?

I wonder if we were to look at the callers of this library if we can just use the pipe natively there, with some refactoring. And just remove this lib. Then the caller of this lib would pass the writer to one side and the reader to the other, instead of having a two-sided object like the ReadWriteCloser.

**Comment by bfosberry** [comment]
ooh

So turns out you don't need to close the reader, which means we can just implement pipe if we want to. I'll see how crazy that would be to do, but for now I'll keep this around

**Comment by bfosberry** [comment]
https://github.com/codeship/go-concurrent/pull/3

**Comment by ngauthier** [comment]
So then can you just embed them so we don't need delegation methods?

**Comment by bfosberry** [comment]
We CAN embed, but its simpler to just update this dep for now.

**Comment by ngauthier** [comment]
ok

**Pull request merged**